galaxyproject/galaxy
# Client side logging

This re-tools the client-side metrics/logging stuff to allow developers (and end users) to turn console logging on and off, have it persist across sessions, and capture logging at page start (to show initialization). The use-case is a developer/user who wants to review logs of what's occurring in the browser during an error or operation from multiple sources without having to add watch statements for each.

To test:
1. open the galaxy homepage in a browser and open the javascript console
2. enter: `Galaxy.debugging( true )`
3. refresh the page
-> You should see a large amount of logging messages prefixed by their namespaces ('GalaxyApp', 'history', 'collection', etc.)

To focus/filter on a particular namespace:
1. in the console, enter: `Galaxy.debuggingNamespaces([ 'collections', 'GalaxyApp' ])` 
-> Only logging from the collections and GalaxyApp namespaces are shown. Hopefully, this allows a decrease in noise v. signal.

To **turn off namespace** filtering: `Galaxy.debuggingNamespaces( null )`
To **turn off debugging**: `Galaxy.debugging( false )`
(You'll need to refresh the page for these to take effect usually)

These settings use localStorage and are persistent across sessions and tabs/windows.

To add namespacing and logger messages to your javascript objects:
- Extend MVC objects from base-mvc.LoggableMixin, call add-logging.addLogging( obj, namespace ) for other objects
- Call this.log, this.debug, this.warn, etc. from within methods
- I've found broader namespaces (per mvc directory) to be useful so far

To be improved at some point during the current epoch:
- **filename and lineno are lost** due to limited console flexibility across browsers
- namespacing could be made easier (again a lack of __filename)

Tested in: Firefox, Chrome, Opera, Safari, Win7+IE11, Win10+Edge

@guerler
